glean-core/ios/Glean/Net/HttpPingUploader.swift

```diff
@@ -150,36 +150,33 @@ public class HttpPingUploader {
     func process() {
         // Limits are enforced by glean-core to avoid an inifinite loop here.
         // Whenever a limit is reached, this binding will receive `.done` and step out.
-        while true {
-            let task = gleanGetUploadTask()
-
-            switch task {
-            case let .upload(request):
-                var body = Data(capacity: request.body.count)
-                body.append(contentsOf: request.body)
-                self.upload(path: request.path, data: body, headers: request.headers) { result in
-                    let action = gleanProcessPingUploadResponse(request.documentId, result)
-                    switch action {
-                    case .next:
-                        // launch a new iteration.
-                        Dispatchers.shared.launchAsync {
-                            HttpPingUploader(configuration: self.config, testingMode: self.testingMode).process()
-                        }
-                    case .end:
-                        return
+        let task = gleanGetUploadTask()
+
+        switch task {
+        case let .upload(request):
+            var body = Data(capacity: request.body.count)
+            body.append(contentsOf: request.body)
+            self.upload(path: request.path, data: body, headers: request.headers) { result in
+                let action = gleanProcessPingUploadResponse(request.documentId, result)
+                switch action {
+                case .next:
+                    // launch a new iteration.
+                    Dispatchers.shared.launchAsync {
+                        HttpPingUploader(configuration: self.config, testingMode: self.testingMode).process()
                     }
-
+                case .end:
+                    return
                 }
 
-                // we don't want to launch multiple uploads at once.
-                // if the upload finishes, we going to launch the next one.
-                return
-            case .wait(let time):
-                sleep(UInt32(time) / 1000)
-                continue
-            case .done:
-                return
             }
+        case .wait(let time):
```

Comment:
The change makes sense and it makes sense that how it was set up before could've caused multiple processes to run at once.

I don't know a lot about this, so this is more of a question for me. Is the only instance when we need to spawn a new process when we hit `.wait`? Just based on the verbiage and the cases I was curious why we wouldn't create a new process on `.done`. Is `.done` just more of a teardown state?

Reply:
`.done` in this case means there are no more pings to send according to glean-core so no need to spawn another `process()`.  The only times this should spawn a new `process()` to handle the next ping is either when the previous ping response is complete or after glean-core has told us to back off (`.wait`).

Reply:
Ok cool, thanks! The changes make sense, but I don't think I have enough context/expertise to give it a proper sign-off.